src/codegen.cpp

```diff
@@ -4580,17 +4589,66 @@ static void allocate_gc_frame(jl_codectx_t &ctx, BasicBlock *b0)
 
     // allocate a placeholder gc instruction
     ctx.ptlsStates = ctx.builder.CreateCall(prepare_call(jltls_states_func));
-    int nthfield = offsetof(jl_tls_states_t, safepoint) / sizeof(void*);
-    ctx.signalPage = emit_nthptr_recast(ctx, ctx.ptlsStates, nthfield, tbaa_const,
-                                        PointerType::get(T_psize, 0));
+
+    Value *ptls_pv = emit_bitcast(ctx, ctx.ptlsStates, T_pprjlvalue);
+    const int ct_offset = offsetof(jl_tls_states_t, current_task);
+    Value *pct =
+        ctx.builder.CreateInBoundsGEP(ptls_pv,
+                                      ConstantInt::get(T_size, ct_offset / sizeof(void *)),
+                                      "current_task_field");
+    LoadInst *ct =
+        ctx.builder.CreateAlignedLoad(pct, Align(sizeof(void *)), "current_task");
+    tbaa_decorate(tbaa_const, ct);
+    ctx.currentTask = ct;
 }
 
+// Store world age at the entry block of the function. This function should be
+// called right after `allocate_gc_frame` and there should be no context switch.
 static void emit_last_age_field(jl_codectx_t &ctx)
 {
+    auto ptls = ctx.ptlsStates;
+    assert(ctx.builder.GetInsertBlock() == ptls->getParent());
     ctx.world_age_field = ctx.builder.CreateInBoundsGEP(
             T_size,
-            ctx.builder.CreateBitCast(ctx.ptlsStates, T_psize),
-            ConstantInt::get(T_size, offsetof(jl_tls_states_t, world_age) / sizeof(size_t)));
+            ctx.builder.CreateBitCast(ptls, T_psize),
+            ConstantInt::get(T_size, offsetof(jl_tls_states_t, world_age) / sizeof(size_t)),
+            "world_age");
+}
+
+// Get current task.
+static Instruction *get_current_task(jl_codectx_t &ctx)
+{
+    return ctx.currentTask;
+}
+
+// Get PTLS through current task.
+static Instruction *get_current_ptls(jl_codectx_t &ctx)
+{
+    Value *task_pv =
+        emit_bitcast(ctx, decay_derived(ctx, get_current_task(ctx)), T_pprjlvalue);
+    const int ptls_offset = offsetof(jl_task_t, ptls);
+    Value *pptls = ctx.builder.CreateInBoundsGEP(
+        T_prjlvalue, task_pv, ConstantInt::get(T_size, ptls_offset / sizeof(void *)),
+        "ptls_field");
+    LoadInst *ptls_load = ctx.builder.CreateAlignedLoad(
+        emit_bitcast(ctx, pptls, T_ppjlvalue), Align(sizeof(void *)), "ptls_load");
+    // Note: Corersponding store (`t->ptls = ptls`) happes in `ctx_switch` of tasks.c.
+    ptls_load->setOrdering(AtomicOrdering::Monotonic);  // TODO: what should we use?
```

Comment:
What's the appropriate ordering here?

Reply:
Private-thread-locals have no atomic ordering

Reply:
Maybe I should've asked how do we prevent a load of PTLS from the field of the task struct to be reordered against a function call? I used atomic ordering so that `b` in this function `f` would not be CSE'ed with `a`:

```julia
function f()
    a = Threads.threadid()
    yield()
    b = Threads.threadid()
    (a, b)
end
```

Is there other ways to do it? Also, since task struct would be shared between threads once we start migration, I thought some kind of ordering might make sense.

Reply:
That's an aliasing question and has nothing to do with atomics